WalletWasabi.Tests/UnitTests/BlockchainAnalysis/PubKeyReuseAnonScoreTests.cs

```diff
@@ -0,0 +1,154 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using WalletWasabi.Blockchain.Keys;
+using WalletWasabi.Tests.Helpers;
+using Xunit;
+
+namespace WalletWasabi.Tests.UnitTests.BlockchainAnalysis
+{
+	public class PubKeyReuseAnonScoreTests
+	{
+		[Fact]
+		public void AddressReusePunishment()
+		{
+			// If there's reuse in input and output side, then output side didn't gain, nor lose anonymity.
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var km = ServiceFactory.CreateKeyManager();
+			var reuse = BitcoinFactory.CreateHdPubKey(km);
+			var tx = BitcoinFactory.CreateSmartTransaction(
+				9,
+				Enumerable.Repeat(Money.Coins(1m), 9),
+				new[] { (Money.Coins(1.1m), 100, BitcoinFactory.CreateHdPubKey(km)) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, reuse) });
+
+			// Make the reused key anonymity set something smaller than 109 (which should be the final anonymity set)
+			reuse.AnonymitySet = 30;
+
+			analyser.Analyze(tx);
+
+			Assert.All(tx.WalletInputs, x => Assert.Equal(100, x.HdPubKey.AnonymitySet));
+
+			// It should be smaller than 30, because reuse also gets punishment.
```

Comment:
should `reuse.AnonymitySet < 30` also be asserted?

Reply:
That is what asserted: `tx.WalletOutputs.First().HdPubKey` is `reuse`. If you want I can remove `tx.WalletOutputs.First().HdPubKey` and reference `reuse` instead.